### PR TITLE
Generate OTX cli_report.log in output folder (otx train only)

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

Add the function to generate a report that bundles multiple summaries while dumping the config used by the task to otx train.
image

The contents of `cli_report.log` are shown below.
```

 ██████╗     ████████╗    ██╗  ██╗
██╔═══██╗    ╚══██╔══╝    ╚██╗██╔╝
██║   ██║       ██║        ╚███╔╝
██║   ██║       ██║        ██╔██╗
╚██████╔╝       ██║       ██╔╝ ██╗
 ╚═════╝        ╚═╝       ╚═╝  ╚═╝

------------------------------------------------------------

Current path: /home/harimkan/workspace/repo/otx-fork/otx-workspace-CLASSIFICATION
sys.argv: ['otx train']
OTX: 1.2.0rc0
------------------------------------------------------------

Running Environments

------------------------------------------------------------
	sys.platform: linux
	Python: 3.9.13 (main, Aug 25 2022, 23:26:10) [GCC 11.2.0]
	CUDA available: True
	GPU 0,1: NVIDIA GeForce RTX 3090
	CUDA_HOME: /usr/local/cuda
	NVCC: Cuda compilation tools, release 11.7, V11.7.64
	GCC: gcc (Ubuntu 9.5.0-1ubuntu1~22.04) 9.5.0
	PyTorch: 1.13.1+cu117
	TorchVision: 0.14.1+cu117
	OpenCV: 4.7.0
	MMCV: 1.7.0
	MMCV Compiler: GCC 9.3
	MMCV CUDA Compiler: 11.7
------------------------------------------------------------

Template Information

------------------------------------------------------------
	model_template_id: 'Custom_Image_Classification_EfficinetNet-B0'
	model_template_path: 'template.yaml'
	name: 'EfficientNet-B0'
	task_family: <TaskFamily.VISION: 1>
	task_type: CLASSIFICATION
	instantiation: <InstantiationType.CLASS: 2>
	summary: 'Class-Incremental Image Classification for EfficientNet-B0'
	framework: 'OTEClassification v1.2.3'
	max_nodes: 1
	application: None
	dependencies: []
	initial_weights: None
	training_targets: [<TargetDevice.GPU: 3>, <TargetDevice.CPU: 2>]
	inference_targets: []
	dataset_requirements: DatasetRequirements(classes=None)
	model_optimization_methods: []
	hyper_parameters: HyperParameterData(base_path='./configuration.yaml', parameter_overrides={'learning_parameters': {'batch_size': {'default_value': 64, 'auto_hpo_state': 'POSSIBLE'}, 'learning_rate': {'default_value': 0.0049, 'auto_hpo_state': 'POSSIBLE'}, 'learning_rate_warmup_iters': {'default_value': 0}, 'num_iters': {'default_value': 90}}, 'nncf_optimization': {'enable_quantization': {'default_value': True}, 'enable_pruning': {'default_value': False}, 'pruning_supported': {'default_value': True}, 'maximal_accuracy_degradation': {'default_value': 1.0}}, 'algo_backend': {'train_type': {'default_value': 'Incremental'}}})
	is_trainable: True
	capabilities: ['compute_representations']
	grpc_address: None
	entrypoints: EntryPoints(base='otx.algorithms.classification.tasks.ClassificationTrainTask', openvino='otx.algorithms.classification.tasks.ClassificationOpenVINOTask', nncf='otx.algorithms.classification.tasks.nncf.ClassificationNNCFTask')
	base_model_path: ''
	exportable_code_paths: ExportableCodePaths(default=None, openvino=None)
	task_type_sort_priority: -1
	gigaflops: 0.81
	size: 4.09
	hpo: None
------------------------------------------------------------

Dataset Information

------------------------------------------------------------
train_subset:
	data_root: /home/harimkan/workspace/datasets/cifar10/initial_data/train
val_subset:
	data_root: /home/harimkan/workspace/datasets/cifar10/initial_data/val
test_subset:
	data_root: /home/harimkan/workspace/datasets/cifar10/initial_data/test
------------------------------------------------------------

Configurations

------------------------------------------------------------
model: {'backbone': {'pretrained': True,
              'type': 'otx.OTXEfficientNet',
              'version': 'b0'},
 'head': {'in_channels': -1,
          'loss': {'loss_weight': 1.0, 'type': 'CrossEntropyLoss'},
          'num_classes': 1000,
          'topk': (1, 5),
          'type': 'CustomLinearClsHead'},
 'hierarchical': False,
 'multilabel': False,
 'neck': {'type': 'GlobalAveragePooling'},
 'pretrained': None,
 'task': 'classification',
 'type': 'SAMImageClassifier'}
dist_params: {'backend': 'nccl', 'linear_scale_lr': True}
cudnn_benchmark: True
seed: 5
deterministic: False
hparams: {'dummy': 0}
task_adapt: {'op': 'REPLACE', 'type': 'mpa'}
log_level: 'INFO'
optimizer: {'lr': 0.0049, 'momentum': 0.9, 'type': 'SGD', 'weight_decay': 0.0005}
optimizer_config: {'grad_clip': None, 'type': 'SAMOptimizerHook'}
runner: {'max_epochs': 90, 'type': 'EpochRunnerWithCancel'}
workflow: [('train', 1)]
lr_config: {'min_lr_ratio': 0.0001,
 'policy': 'CosineAnnealing',
 'warmup': None,
 'warmup_iters': 0}
evaluation: {'metric': ['accuracy', 'class_accuracy']}
load_from: None
resume_from: None
checkpoint_config: {'interval': 1, 'max_keep_ckpts': 1, 'type': 'CheckpointHookWithValResults'}
fp16: {'loss_scale': 512.0}
train_type: 'Incremental'
data: {'samples_per_gpu': 64,
 'test': {'domain': <Domain.CLASSIFICATION: 2>,
          'empty_label': None,
          'labels': None,
          'otx_dataset': None,
          'pipeline': [{'size': 224, 'type': 'Resize'},
                       {'mean': [123.675, 116.28, 103.53],
                        'std': [58.395, 57.12, 57.375],
                        'to_rgb': False,
                        'type': 'Normalize'},
                       {'keys': ['img'], 'type': 'ImageToTensor'},
                       {'keys': ['img'], 'type': 'Collect'}],
          'test_mode': True,
          'type': 'OTXClsDataset'},
 'test_dataloader': {},
 'train': {'domain': <Domain.CLASSIFICATION: 2>,
           'empty_label': None,
           'labels': None,
           'otx_dataset': None,
           'pipeline': [{'size': 224, 'type': 'Resize'},
                        {'direction': 'horizontal',
                         'flip_prob': 0.5,
                         'type': 'RandomFlip'},
                        {'config_str': 'augmix-m5-w3', 'type': 'AugMixAugment'},
                        {'angle': (-10, 10), 'p': 0.35, 'type': 'RandomRotate'},
                        {'keys': ['img'], 'type': 'PILImageToNDArray'},
                        {'mean': [123.675, 116.28, 103.53],
                         'std': [58.395, 57.12, 57.375],
                         'to_rgb': False,
                         'type': 'Normalize'},
                        {'keys': ['img'], 'type': 'ImageToTensor'},
                        {'keys': ['gt_label'], 'type': 'ToTensor'},
                        {'keys': ['img', 'gt_label'],
                         'meta_keys': {'filename',
                                       'flip',
                                       'flip_direction',
                                       'ignored_labels',
                                       'img_norm_cfg',
                                       'img_shape',
                                       'ori_filename',
                                       'ori_shape',
                                       'pad_shape',
                                       'scale_factor'},
                         'type': 'Collect'}],
           'type': 'OTXClsDataset'},
 'train_dataloader': {},
 'val': {'domain': <Domain.CLASSIFICATION: 2>,
         'empty_label': None,
         'labels': None,
         'otx_dataset': None,
         'pipeline': [{'size': 224, 'type': 'Resize'},
                      {'mean': [123.675, 116.28, 103.53],
                       'std': [58.395, 57.12, 57.375],
                       'to_rgb': False,
                       'type': 'Normalize'},
                      {'keys': ['img'], 'type': 'ImageToTensor'},
                      {'keys': ['img'], 'type': 'Collect'}],
         'test_mode': True,
         'type': 'OTXClsDataset'},
 'val_dataloader': {},
 'workers_per_gpu': 2}
early_stop_metric: 'accuracy'
early_stop: {'iteration_patience': 0, 'patience': 8, 'start': 3}
custom_hooks: [{'enable_adaptive_interval_hook': False,
  'enable_eval_before_run': True,
  'type': 'AdaptiveTrainSchedulingHook'},
 {'interval': 1,
  'iteration_patience': 0,
  'metric': 'accuracy',
  'patience': 8,
  'priority': 75,
  'start': 3,
  'type': 'LazyEarlyStoppingHook'},
 {'init_callback': 'otx.algorithms.common.tasks.training_base.OnHookInitialized',
  'type': 'CancelInterfaceHook'},
 {'priority': 71,
  'time_monitor': <otx.algorithms.common.utils.callback.TrainingProgressCallback object at 0x7f0c4488b4c0>,
  'type': 'OTXProgressHook',
  'verbose': True},
 {'priority': 'LOWEST', 'type': 'ForceTrainModeHook'},
 {'priority': 'VERY_LOW', 'type': 'MemCacheHook'}]
------------------------------------------------------------

Results

------------------------------------------------------------
	time elapsed: '0:00:13.886423'
	score: Performance(score: 0.7236842105263158, dashboard: (3 metric groups))
	model_path: '/home/harimkan/workspace/repo/otx-fork/otx-workspace-CLASSIFICATION/outputs/20230330_180006_train/models'

```

### How to test

```
(otx) otx train
# unit-test
(otx) pytest -s -v tests/unit/cli/utils/test_report.py
```

### Checklist

- [X] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [x] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [x] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

File: otx/cli/tools/train.py [hunk 1]
```diff
@@ -250,8 +257,23 @@ def main():  # pylint: disable=too-many-branches
...
         task.evaluate(resultset)
-        assert resultset.performance is not None
-        print(resultset.performance)
+        performance = resultset.performance
+        assert performance is not None
+        print(performance)
+
+    end_time = time.time()
+    sec = end_time - start_time
+    total_time = str(datetime.timedelta(seconds=sec))
```
Comment (JihwanEom) [inline otx/cli/tools/train.py:266]: How about just to use `end - start` without datetime? Do we have special reason to use datetime?

File: otx/cli/tools/train.py [hunk 2]
```diff
@@ -250,8 +257,23 @@ def main():  # pylint: disable=too-many-branches
...
+    print("otx train time elapsed: ", total_time)
+    model_results = {"time elapsed": total_time, "score": performance, "model_path": str(model_path.absolute())}
```
Comment (JihwanEom) [inline otx/cli/tools/train.py:268]: Hmm, it's just a question, do you have any plan to include `trained epochs` and `best epoch` information on other PR?

File: otx/cli/utils/report.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+"""Report Generating for OTX CLI."""
```
Comment (JihwanEom) [inline otx/cli/utils/report.py:1]: How about to define a class named `OTXReportGenerator`?

**Review by JihwanEom — COMMENTED** [review]
Thanks for great work! Did you check all tasks works well with OTX report?

**Comment by harimkang** [comment]
> Thanks for great work! Did you check all tasks works well with OTX report?

I only checked classification and anomaly.  Are there any other taks you are concerned about?

**Comment by JihwanEom** [comment]
> > Thanks for great work! Did you check all tasks works well with OTX report?
> 
> I only checked classification and anomaly. Are there any other taks you are concerned about?

All the remained tasks.. :D

**Comment by JihwanEom** [comment]
And why `otx_dataset` is None for above report?

**Comment by harimkang** [comment]
> And why `otx_dataset` is None for above report?

i don't know, if there is a problem, it must be a problem with the current develop task.

File: otx/cli/tools/train.py (hunk 1 above)
Comment (harimkang) [reply otx/cli/tools/train.py:266]: It's just because it's easy to see.

File: otx/cli/tools/train.py (hunk 2 above)
Comment (harimkang) [reply otx/cli/tools/train.py:268]: I don't know what trained epoch means, but I don't plan on doing that right now.

File: otx/cli/utils/report.py (hunk 3 above)
Comment (harimkang) [reply otx/cli/utils/report.py:1]: This feature does not use class.

File: otx/cli/tools/train.py (hunk 2 above)
Comment (harimkang) [reply otx/cli/tools/train.py:268]: > I don't know what trained epoch means, but I don't plan on doing that right now.

As you can see from the code configuration, it is possible by including that information in task.config.

**Review by sungmanc — COMMENTED** [review]
It seems a nice feature to improve UX. Could you also update the CHANELOG.md?

**Comment by harimkang** [comment]
> It seems a nice feature to improve UX. Could you also update the CHANELOG.md?

Thanks for letting us know. updated! https://github.com/openvinotoolkit/training_extensions/pull/1959/commits/39cd74ec9dbd02095849f4009dfab485523ffacf

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,11 +6,11 @@ All notable changes to this project will be documented in this file.
 
 ### New features
 
--
+- Add Generating feature cli_report.log in output for otx training (<https://github.com/openvinotoolkit/training_extensions/pull/1959>)
```
Comment (sungmanc) [inline CHANGELOG.md]:
```suggestion
- Add generating feature cli_report.log in output for otx training (<https://github.com/openvinotoolkit/training_extensions/pull/1959>)
```

**Review by sungmanc — DISMISSED** [review]

**Review by jaegukhyun — APPROVED** [review]
LGTM, by the way do we need to add checking existence of report after training in cli test?

**Review by sungmanc — APPROVED** [review]

**Pull request merged**